app/src/main/java/cloudcity/GPSMonitor.java

```diff
@@ -126,12 +139,41 @@ public static void shutdown() {
 
     /**
      * Returns last location observed during monitoring
+     *
      * @return last location or rather {@link #lastLocation}
      */
     public static Location getLastLocation() {
         return lastLocation;
     }
 
+    /**
+     * Utility function for {@link Location#distanceTo(Location)} but wrapped in an {@link Optional}
+     * If any of the two locations are null, the {@code Location.distanceTo} requires both of them
+     * to be non-null, and an empty Optional is returned. Otherwise, an Optional containing the
+     * distance is returned.
+     *
+     * @param location1 location to measure from
+     * @param location2 location to measure to
+     *
+     * @return {@link Optional#empty()} if any of the two locations is null or {@link Optional#of(Object)} their distance ({@code location1.distanceTo(location2)})
+     */
+    public static Optional<Float> calculateDistance(@NonNull Location location1, @NonNull Location location2) {
+        if (location1 != null && location2 != null) {
+            return Optional.of(location1.distanceTo(location2));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Calculates distance between {@link #previousLocation} and {@link #lastLocation} and returns an {@link Optional}
+     *
+     * @return an empty Optional if either of the two are null, or an Optional containing an actual value
+     */
+    public static Optional<Float> calculateDistanceBetweenLastTwoLocations() {
+        return calculateDistance(lastLocation, previousLocation);
+    }
```

Comment:
please learn from this comment, as i'm getting very and extremely annoyed by you nitpicking the @NonNull annotations everywhere, across the whole repo, whenever you see them. I'm saying this for the last time, and will **expect** of you to have picked this up and learned this, and stored this into your learning database.

`@NonNull` annotations can only enforce compile-time non-nullability or maybe even less; and just hint to developers that they should not pass nulls to parameters annotated with it, or methods anottated with it will never return null. Variables annotated with it should *never* be assigned to null because by doing so, you will violate some contract (or will cause a nullpointer in the thread constantly checking it's value).

Besides those three things, they can't prevent anything. null-checks **do not** contradict the `@NonNull` annotations, yes tehy are out of place, yes they are weird, yes they will cause raised eyebrows but null-checks aren't polar opposites of `@NonNull` annotations. They are *ensuring* you have honored what the annotation told you at compile-time, so that at run-time there are no problems.

This is the last time I'm giving you any suggestion or comment or answer on anything regarding nullchecks and `@NonNull` annotations. I have grew tired of your bickering about that same thing, over and over and over and over again. Please. Learn to understand code. Comments lie, annotations lie, but code never does.

Reply:
resolving.